[russell.yoo] 방탈출 예약 대기 관리 Step 1, 2  제출 합니다.

russell-yoo opened the pull request:
마리벨 안녕하세요!😄

이번 미션의 기능 요구사항은 크게 아래 두가지 입니다.

- `auth` 패키지와 `nextstep`패키지의 분리
- `reservation-waitings`의 create, find, delete 구현

reservation-waitings의 schema는 아래와 같습니다.

```sql
CREATE TABLE reservation_waiting
(
    id          bigint not null auto_increment,
    schedule_id bigint not null COMMENT 'SCHEDULE PK(FK)',
    member_id   bigint not null COMMENT 'MEMBER PK(FK)',
    wait_num    bigint not null COMMENT '예약 대기 번호'
);
```

`reservation-waiting`은 해당 `schedule`의 reservation이 존재할 때, `reservaion-waiting`을 생성하고, 존재하지 않으면 `reservation`을 생성하는 로직을 갖고 있습니다. 

또한, 예약 대기 번호는 현재 로직 기준으로 예약 대기번호의 max를 구하고 해당 max + 1 하는 형식으로 구현돼 현재 남은 사람을 구하는 것이 아니라, 고정 된 번호를 이용하여 대기번호를 구현하였습니다. 여기서, 궁금한 사항이 있습니다.

대기번호를 구현하는데 떠오른 방법은 2가지 있습니다. 처음으로 위의 구현 방법과 같이 고정된 대기 번호를 사용하는 방법과, 시간을 기록하여 해당 시간보다 빠른 사람 수를 구하여 자신이 몇 team을 기다려야 하는지 매번 db와 통신하여 구해주는 방법이 있다고 생각합니다.
전자의 방법은 성능에서 이점이 있고, 후자의 방법은 예약 대기 취소가 발생했을 때 효과적이라 생각합니다. 만약 이러한 상황일 때 어느쪽의 구현방법이 더 좋은지 궁금합니다.

매번 꼼꼼하게 리뷰해주셔서 감사합니다!!👏🏻👏🏻👏🏻👏🏻👏🏻👏🏻👏🏻👏🏻👏🏻👏🏻

kakao-jay-den committed:
refactor : auth 패키지 분리

- [x] Auth 패키지를 nextstep 외부로 이동
패키지 분리 후 WebMvcConfiguration 에서 componentScan 으로 auth 에 있는 컴포넌트 등록

kakao-jay-den committed:
refactor : member 의 의존성 제거 - 중간 객체 사용

- [x] member 의 의존성 제거 - 중간 객체 사용
member 대신 UserDetails 를 의존하게 하여 패키지 간 member 의 의존성 제거

kakao-jay-den committed:
fix : member 의 의존성 제거 - 중간 interface 적용

- [x] member 의 의존성 제거 - 중간 객체 사용
MemberDetails 를 interface 로 사용

kakao-jay-den committed:
refactor : member dao 의존성 제거

- [x] memberDao 의 의존성 제거
LoginMemberDao 를 통한 의존성 제거

kakao-jay-den committed:
refactor :  Configuration, Import 사용해서 auth 패키지 내 컴포넌트들을 bean 으로 등록

- [x] Configuration, Import 사용해서 auth 패키지 내 컴포넌트들을 bean 으로 등록

kakao-jay-den committed:
feat : reservation-waiting table scheme 작성

- ReservationWaiting class 작성
- schema.sql 수정

kakao-jay-den committed:
feat : reservation-waiting repository 구현

- ReservationWaitingDao 구현
- ReservationWaitingDao Test Code 작성

kakao-jay-den committed:
feat : reservation-waiting service 구현

- findAll, create, delete 구현

kakao-jay-den committed:
feat : reservation-waiting controller 구현

- create, findByMemberId, deleteById 구현

kakao-jay-den committed:
feat : 통합테스트 작성

- [x] 통합테스트 작성

kakao-jay-den committed:
feat : 통합테스트 작성

- [x] 통합테스트 작성

russell-yoo committed:
refactor: auth package final 적용

russell-yoo committed:
refactor: next-step package final 추가

russell-yoo committed:
refactor: 사용하지 않는 메소드 제거 및 context 테스트 추가"

russell-yoo committed:
refactor: auth package package 분리

russell-yoo committed:
refactor: auth package jsonCreator 정리 및 throw 정리

russell-yoo committed:
refactor: exception logic 변경 및 Member 리팩토링

- exception을 BaseWebException을 상속하는 형태로 변경
- null check를 위한 Validator 구현
- JsonCreator 대신 MemberResponse 사용
- Member Builder 구현 및 MemberResponse Builder 구현

russell-yoo committed:
style: optimize imports

russell-yoo committed:
fix: BaseWebException의 Type인자 로직 변경

russell-yoo committed:
refactor: baseWebException 인자 Class<?> -> String으로 변경

russell-yoo committed:
refactor: reservation package refactoring

- Builder 적용
- Validator 적용
- exception 로직 변경 및 exception 추가

russell-yoo committed:
refactor: ReservationWaiting package refactoring

- Builder 적용
- Validator 적용

russell-yoo committed:
refactor: Schedule package refactoring

- Builder 적용
- Validator 적용
- Response Dto 작성

russell-yoo committed:
refactor: Theme package refactoring

- Builder 적용
- Validator 적용
- Response Dto 작성
- Imports 최적화

russell-yoo committed:
refactor: Exception Handler package 분리

russell-yoo committed:
refactor: validate 로직 통일

russell-yoo committed:
refactor: DAO와 SQL 분리

russell-yoo committed:
refactor: schema comment 추가

russell-yoo committed:
refactor: dao interface 분리

russell-yoo committed:
refactor:  test auth package 분리

russell-yoo committed:
refactor: auth 관련 exception 처리 로직 변경

- auth 관련 exception 분리 및 message 정보를 갖게 변경

russell-yoo committed:
refactor: reservation-waitings 생성로직 개선

Hyunji-Yoo commented on src/main/java/nextstep/member/MemberDao.java:
```diff
@@ -1,55 +1,11 @@
+import auth.config.LoginMemberDao;
 
-import java.sql.PreparedStatement;
+public interface MemberDao extends LoginMemberDao {
```
보다 DIP 스럽네요 👍

russell-yoo replied:
감사합니다.

Hyunji-Yoo commented on src/main/java/nextstep/reservationwaiting/ReservationWaitingService.java:
```diff
@@ -0,0 +1,55 @@
+        return reservationWaitingDao.save(newReservationWaiting);
+    }
+
+    public List<ReservationWaiting> findAllByMemberId(Member member) {
```
메서드명이랑 파라미터가 매칭이 잘 안되네요.
ByMemberId라면 파라미터로 memberId만 받는게 맞을 것 같아요.
아니면 반대로 member를 받아도 어색하지 않을 범용적 이름으로 메서드명을 변경해도 될 것 같구요.
파라미터는 반드시 필요한 아규먼트만 받도록 정의하는게 좋습니다.
반드시 필요한 아규먼트가 여러개인 경우에는 파라미터를 객체로 받도록 하는게 낫구요.

Hyunji-Yoo commented on src/main/java/nextstep/reservationwaiting/ReservationWaitingService.java:
```diff
@@ -0,0 +1,55 @@
+        this.scheduleDao = scheduleDao;
+    }
+
+    public Long create(Member member, ReservationWaitingRequest reservationWaitingRequest) {
```
ℹ️ NIT
실무에서는 로그인한 사용자정보는 일반적으로 세션에 담아두고 필요한 경우 서비스에서 그걸 가져와서 사용해요.
그래서 reservationWaitingRequest만 받으면 된답니다.

Hyunji-Yoo commented on src/main/java/nextstep/reservationwaiting/ReservationWaitingController.java:
```diff
@@ -0,0 +1,57 @@
+        boolean isDuplicate = reservationService.isDuplicateByScheduleId(reservationWaitingRequest.getScheduleId());
+        if (isDuplicate) {
+            Long id = reservationWaitingService.create(member, reservationWaitingRequest);
+            return ResponseEntity.created(URI.create("/reservation-waitings/" + id)).build();
```
갑자기 궁금한건데
reservationWaiting 생성 후 응답값을 `"/reservation-waitings/" + id` URI로 리턴하는 이유가 뭔가요?
아마 생성한 reservationWaiting를 조회할 수 있도록 조회 API URI를 넘겨주는건가 싶은데
ReservationWaitingController 내에서 `"/reservation-waitings/" + id`에 매핑되는 API는 삭제 API 뿐인 것 같아서요ㅎㅎ

russell-yoo replied:
기능 요구사항에 해당하는 내용이 적혀 있어 구현하였는데 해당 api에 매핑되는 것이 없네요.

그래서 해당하는 로직을 구현하였고, 해당 로직에는 현재 대기자 수를 추가하였습니다.

Hyunji-Yoo commented on src/main/java/nextstep/reservationwaiting/ReservationWaitingController.java:
```diff
@@ -0,0 +1,57 @@
+            Long id = reservationWaitingService.create(member, reservationWaitingRequest);
+            return ResponseEntity.created(URI.create("/reservation-waitings/" + id)).build();
+        }
+        Long id = reservationService.create(member, new ReservationRequest(reservationWaitingRequest.getScheduleId()));
```
reservation 생성을 reservation-waiting 컨트롤러에서 하는건 SRP에 위반하는 것 같아요.
아니면 여기서 reservation 컨트롤러로 바로 리다이렉트 시키는게 맞을 것 같아요.
혹은 예외를 발생시키고 그 예외를 처리하는 AOP에서 리다이렉트 URL을 반환하고 웹(FE)에서 그걸 받아서 리다이렉트 시킬 수도 있겠죠.

russell-yoo replied:
여기서 질문이 있습니다.

현재 `exception`을 발생시키고, 해당 `exception`가 발생했을 때 `exception handler`를 통해 따로 생성해주는 로직으로 구현하였는데 `exception handler`가 `reservationWaitingController`에 위치하고 있습니다. 

`exception handler`를 따로 관리하려면 어떠한 방법을 통해 관리하는게 좋을지 궁금하여 질문드렸습니다. 

기존의 `AuthExceptionHandler`와 `BaseExceptionHandler`가 있었는데 이 두가지는 영어 사전순으로 `exception handler`가 호출돼 문제가 없었으나, `ReservationWaitingExceptionHandler`를 추가하여 구현하게 되면 `BaseExceptionHandler`의 `Exception`을 관리하는 `onException`에 영향을 받아 제대로 `Exception`이 관리가 되지 않음을 확인했습니다. 이에 따라 `AuthExceptionHandler`와 `BaseExceptionHandler` 를 하나의 파일로 바꾸었고, `ReservationWaitingContrller`에  생성해주는 handler를 위치시켜 해결했습니다.

RestControllerAdvice를 쓰기에는 공통된 범위를 갖고 있고,  ABC순으로 배치하는 것은 나중에 관리에 용이하지 않다고 생각합니다.  또한, Controller에 Exception Handler를 배치하는 것이 옳은지 궁금하고, 실제로는 어떠한 방식으로 관리하는게 좋은지 궁금하여 질문드립니다.

Hyunji-Yoo replied:
@yooj929 
**1. `ControllerAdvice`가 여러개인 경우 적용순서에 따라 발생하는 이슈**
> 우선 해당 이슈는 `@Order` 어노테이션을 사용해 해결 할 수 있을 것 같습니다.
> 
> ```
> @ControllerAdvice
> @Order(Ordered.HIGHEST_PRECEDENCE)
> public class AuthExceptionHandler {
> ```
> ```
> @ControllerAdvice
> @Order(Ordered.LOWEST_PRECEDENCE)
> public class BaseExceptionHandler {
> ```
> 이런 식으로 하면 `BaseExceptionHandler.onException`의 적용순서를 후순위로 설정할 수 있지 않을까 싶어요.


**2. Controller에 `ExceptionHandler`를 배치하는 것이 옳은지 궁금하고, 실제로는 어떠한 방식으로 관리하는게 좋은지**
> 이건 프로젝트의 규모나 서비스의 특성에 따라 달라지는 것이지 더 나은 방안이 정해져 있다고 보기는 어려울 것 같아요.
> 
> 제공하는 서비스가 다양해서 Controller가 많다면 ExceptionHandler들을 따로 모아서 ControllerAdvice로 분리하는게 관리하기 훨씬 용이하겠죠.
> 
> 하지만 작은 서비스이고 특정 컨트롤러에서만 발생하는게 명확한 Exception이거나 
> 동일한 Exception이어도 컨트롤러 별로 처리하는 방법이 달라야 한다면
> Controller 내에 ExceptionHandler를 직접 배치하는 것도 나쁘지 않다고 생각합니다.
> 
> 제 개인적인 선호는 ControllerAdvice로 분리해서 가급적 모듈화를 할 것 같네요.

russell-yoo replied:
감사합니다

Hyunji-Yoo reviewed (changes requested):
안녕하세요 러셀.

대기번호까지 생기고 꽤 규모있는 서비스가 되어 가고 있군요 멋지네요 👍
고민을 더 해보면 좋을 것 같은 부분에 코멘트 달았습니다. 
반영 후 리뷰 재요청 주시면 확인할게요 😉

항상 응원합니다.

Hyunji-Yoo commented:
추가로 문의주신 내용과 관련해 지극히 개인적인 저의 의견을 드립니다.

일단 구조에서 대기번호를 구현한다면 러셀이 설계하신 것처럼 고정값 max로 대기번호를 발행하는게 맞을 것 같아요.

다만 테이블에 status 컬럼을 추가하고 `ReservationWatingStatus`의 enum값을 함께 저장하도록 하겠죠.
```
- WAITING(1)
- IN_PROGRESS(2)
- FINISHED(3)
- DROPPED(4)
```

발행 시에 해당 상태에 대한 조회도 추가해야겠죠.
```
SELECT	MAX(wait_num)
FROM	reservation_waiting
WHERE	schedule_id = ${예약하려는_스케줄번호}
AND	status in (1, 2)
```

사용자가 대기번호를 받고 대기번호조회화면으로 이동하면 그때는 status 3이고, 나의 wait_num보다 큰 개수를 구해서 "대기자 수"를 보여줄 수 있겠네요.
해당 화면에 새로고침 버튼을 둬서 누를 때마다 재조회하게 하던지 아니면 클라이언트에서 폴링을 하게 하거나, 그것도 아니면 웹소켓으로 푸시받도록 할 수 있겠네요.
```
SELECT	COUNT(id)
FROM	reservation_waiting
WHERE	schedule_id = ${예약하려는_스케줄번호}
AND	status = 3
AND	wait_num > ${나의_순번}
```
또한 해당 화면에서 대기취소를 하는 기능도 제공하면 좋겠어요. 그 경우에는 status를 DROPPED로 변경하구요.

reservation을 생성할 때는 우선 reservation_wating의 상태를 `IN_PROGRESS`로 바꾸고 
reservation 생성이 완료되면 reservation_wating의 상태를 `FINISHED`로 바꿔야되겠죠.

추가로 modified_at 컬럼을 추가해서 expired 된 대기를 찾아서 종료처리하는 기능도 있을 수 있겠어요.

사실 실무에서라면 따로 대기번호 발행시스템을 구축할 것 같아요 😉
그때는 DB 기반으로 하기 보다는 redis 등의 미들웨어를 이용할거에요!

부족한 설명이지만 도움이 좀 됐으면 좋겠네요ㅎㅎ
그럼 화이팅하세요!

russell-yoo committed:
refactor: ReservationWaitingService의 signature 변경

russell-yoo committed:
refactor: ReservationWaitingController의 로직 변경

- exception handler로 로직 이관
- exception handler 버그 수정

russell-yoo committed:
feat: reservationWaitingDao findById 구현

russell-yoo committed:
feat: reservationWaitings 단건 조회 구현

russell-yoo committed:
feat: 현재 대기자수 구하는 로직 추가

- reservation_waiting table에 wait_status 추가
- InvalidFieldException, InvalidChangeStatusException 추가
- exception handler 버그 수정
- reservation-waitings 삭제시 상태 추가에 따른 admin/ user 기능 분리
- 현재 대기자수 구하는 로직 추가
- response에 status 필드 추가

yooj929 commented:
reservation_waiting의 schema를 다양한 상황에 따라 처리하기 위해 제안해주신 방법으로 변경 하였습니다.
*As-is*
```sql

CREATE TABLE reservation_waiting
(
    id          bigint not null auto_increment,
    schedule_id bigint not null COMMENT 'SCHEDULE PK(FK)',
    member_id   bigint not null COMMENT 'MEMBER PK(FK)',
    wait_num    bigint not null COMMENT '예약 대기 번호'
);

```

*To-be*
```sql
CREATE TABLE reservation_waiting
(
    id          bigint  not null auto_increment,
    schedule_id bigint  not null COMMENT 'SCHEDULE PK(FK)',
    member_id   bigint  not null COMMENT 'MEMBER PK(FK)',
    wait_num    bigint  not null COMMENT '예약 대기 번호',
    wait_status tinyint not null COMMENT '예약 대기 상태'
);

```
---


```sql
SELECT	MAX(wait_num)
FROM	reservation_waiting
WHERE	schedule_id = ${예약하려는_스케줄번호}
AND	status in (1, 2)
```

위의 기능은 예약 대기를 위한 대기 번호를 조회하는 기능이라고 생각합니다. 
현재 대기 번호는 예약 취소가 빈번하게 돼 이를 방지하기 위한 기능으로 현재 대기 번호는 몇 팀이 취소되면 자신이 이용할 수 있는가를 의미 하고 있습니다. 그래서 AND status in (3)으로 구현 하였고, 상태를 세부적으로 나누어 개발하게 된다면 여러 상황에 대처할 수 있는 것을 알게 돼 좋았습니다.😄
그 예로 해당 상태를 보고 삭제 로직을 `admin`은 `reservation_waiting` table의 row를 지우고, user는 `status`를 `DROPPED`로 변경하는 형식으로 구현하였고, 이렇게 구현하게 되면 abuser 탐색이나, 자주 취소하는 사람에 대한 관리를 할 수 있을 것이라 생각됩니다.

auth package분리에 대해서 하나 궁금증이 생겼습니다.
현재 `MemberDetails` 인터페이스를`Member`가 상속받고`argumentResolver`가 `MemberDetails`를 return하고 이를 `Member`로 down-casting하는 형식으로 `Member`는 `MemberDetails`를 의존하고 `argumentResolver`는 `MemberDetails`를 의존하는 방법으로 구현하였는데, 해당 방법과 중간에 변환 객체를 이용한 방법 중 어떠한 방법이 더 좋은지 궁금합니다.

감사합니다.😉

Hyunji-Yoo commented on src/main/java/auth/config/LoginMemberDao.java:
```diff
@@ -0,0 +1,6 @@
+package auth.config;
```
`LoginMemberDao, MemberDetails`가 config에 있는게 좀 의아하긴 하네요.
config에는 정말 구성/설정 관련된 모듈만 배치하는게 맞을 것 같아요.
nextstep 패키지와 유사하게 auth 패키지 하위에 `LoginController` 등과 같이 있는게 위상이 맞지 않을까 싶네요.

ℹ️ NIT
패키지명이 auth라서 소스코드명도 Auth*로 가면 더 일관성있을 것 같긴해요

Hyunji-Yoo commented:
> reservation_waiting의 schema를 다양한 상황에 따라 처리하기 위해 제안해주신 방법으로 변경 하였습니다.
> 
> _As-is_
> 
> ```sql
> CREATE TABLE reservation_waiting
> (
>     id          bigint not null auto_increment,
>     schedule_id bigint not null COMMENT 'SCHEDULE PK(FK)',
>     member_id   bigint not null COMMENT 'MEMBER PK(FK)',
>     wait_num    bigint not null COMMENT '예약 대기 번호'
> );
> ```
> 
> _To-be_
> 
> ```sql
> CREATE TABLE reservation_waiting
> (
>     id          bigint  not null auto_increment,
>     schedule_id bigint  not null COMMENT 'SCHEDULE PK(FK)',
>     member_id   bigint  not null COMMENT 'MEMBER PK(FK)',
>     wait_num    bigint  not null COMMENT '예약 대기 번호',
>     wait_status tinyint not null COMMENT '예약 대기 상태'
> );
> ```
> 
> ```sql
> SELECT	MAX(wait_num)
> FROM	reservation_waiting
> WHERE	schedule_id = ${예약하려는_스케줄번호}
> AND	status in (1, 2)
> ```
> 
> 위의 기능은 예약 대기를 위한 대기 번호를 조회하는 기능이라고 생각합니다. 현재 대기 번호는 예약 취소가 빈번하게 돼 이를 방지하기 위한 기능으로 현재 대기 번호는 몇 팀이 취소되면 자신이 이용할 수 있는가를 의미 하고 있습니다. 그래서 AND status in (3)으로 구현 하였고, 상태를 세부적으로 나누어 개발하게 된다면 여러 상황에 대처할 수 있는 것을 알게 돼 좋았습니다.😄 그 예로 해당 상태를 보고 삭제 로직을 `admin`은 `reservation_waiting` table의 row를 지우고, user는 `status`를 `DROPPED`로 변경하는 형식으로 구현하였고, 이렇게 구현하게 되면 abuser 탐색이나, 자주 취소하는 사람에 대한 관리를 할 수 있을 것이라 생각됩니다.
> 
> auth package분리에 대해서 하나 궁금증이 생겼습니다. 현재 `MemberDetails` 인터페이스를`Member`가 상속받고`argumentResolver`가 `MemberDetails`를 return하고 이를 `Member`로 down-casting하는 형식으로 `Member`는 `MemberDetails`를 의존하고 `argumentResolver`는 `MemberDetails`를 의존하는 방법으로 구현하였는데, 해당 방법과 중간에 변환 객체를 이용한 방법 중 어떠한 방법이 더 좋은지 궁금합니다.
> 
> 마지막으로 이번 스프링 미션 중 선택 미션을 진행하지 않고 마무리 하려고 합니다. 그 이유는 2월 3일 이후로는 Spring 미션에 대해 리뷰 요청이 불가하고, nextstep에서 WAS 미션을 중점적으로 하라는 연락을 받았습니다.
> 
> 감사합니다.😉

음 저는 지금 구현된 대로 인터페이스에 의존하는 방식을 좋아하긴 하는데...
지금 구조에서는 `MemberDetails`에 의존하는 `Member`가 Entity인데 컨트롤러에서 
`@LoginMember Member member` 이렇게 바로 해당 Entity를 Request에서 사용하는게 바람직하지는 않을 것 같아요.
LoginMemberArgumentResolver에서 바로 `Member`를 반환하기 보다는 불변DTO 클래스로 변환하여 반환하는게 더 나을 것 같습니다.

Hyunji-Yoo reviewed (approved):
안녕하세요 러셀.
궁금하신 점 남겨주신 내용에 대해서는 코멘트 남겼습니다 😉
원활한 과제 수행을 위해서 우선 머지하겠습니다.
이번 주도 고생 많으셨어요 👍
